docs/map.md

```diff
@@ -0,0 +1,172 @@
+the `map` components:
+
+```hcl
+globals {
+    map obj {
+        for_each = [
+            {name="a", value=5},
+            {name="c", value=0},
+            {name="a", value=15},
+            {name="b", value=5},
+            {name="c", value=20},
+        ]
+        iterator = elem
+        key = elem.new.name
+        value {
+            value = elem.new.value
+            previous = tm_try(elem.old.value, null)
+        }
+    }
+}
+
+```
+
+The code above will result in the `global.obj` defined below:
+
+```hcl
+obj = {
+  a = {
+    previous = 5
+    value    = 15
+  }
+  b = {
+    previous = null
+    value    = 5
+  }
+  c = {
+    previous = 0
+    value    = 20
+  }
+}
+
+
+```
+
+The `map` will iterate over the value in `for_each`, setting the iterator variable (`elem` in this case) with the **new** and **old** value for the
+provided `key`. If the `iterator` attribute is not set, then the default
+iterator name is `element`. The `element.new` contains the current iterated
+value and the `element.old` contains the previous value stored for the current
+`key`.
+
+Now it's time for a more useful example.
+
+Have a look in the `global.orders` list declared below:
+
+```
+globals {
+    orders = [
+        {name = "Morpheus", product = "sunglass", price = 100.5},
+        {name = "Trinity",  product = "cape", price = 82.30},
+        {name = "Trinity",  product = "necklace", price = 25.0},
+        {name = "Trinity",  product = "sunglass", price = 100.5},
+        {name = "Anderson", product = "ollydbg", price = 30},
+        {name = "Morpheus", product = "boot", price = 65},
+        {name = "Anderson", product = "cape", price = 82.30},
+        {name = "Morpheus", product = "sunglass", price = 145.50},
+    ]
+}
+```
+
+Let's say you would like to aggregate those values by `name`, then you can do it
+with the `map` block as defined below:
+
+```
+map totals {
+        for_each = global.orders
+        key = element.new.name
+        value {
+            total_spent = tm_try(element.old.total_spent, 0) + element.new.price
+        }
+    }
+```
+
+Which will result in the object below:
+
+```
+totals = {
+  Anderson = {
+    total_spent = 112.3
+  }
+  Morpheus = {
+    total_spent = 311
+  }
+  Trinity = {
+    total_spent = 207.8
+  }
+}
+```
+
+But you can aggregate even further by using nested `map` blocks and then
+aggregate by `product` for each `name`. 
+
+Example:
+
+```hcl
+globals {
+    map totals {
+        for_each = global.orders
+        key = element.new.name
+        value {
+            total_spent = tm_try(element.old.total_spent, 0) + element.new.price
+
+            map per_product {
+               for_each = [for v in global.orders : v if v.name == element.new.name]
+               key = element.new.product
+               value {
+                   total = tm_try(element.old.total, 0) + element.new.price
+               }
+            }
```

Comment:
can't we use element.old.per_product somehow?

what is the context of `for_each` here? element.old and element.new are the values of the outer block? 

we should set a second iterator to make this example more readable..

can't we do something like: `for_each = [ .... element.old.per_product... element.new.... ]` as we just want to extend the old per_product map? I will try it out later ;)

Reply:
> can't we use element.old.per_product somehow?

inside the nested map or from outside? From outside it's already supported but the problem is how to merge them. Just did this:

```
globals {
  map totals {
    for_each = global.orders
    key = element.new.name
    value {
      total_spent = tm_try(element.old.total_spent, 0) + element.new.price
      per_product = tm_can(element.old.per_product) ? tm_merge(element.old.per_product, {
              "${element.new.product}" = {
                  total = tm_try(element.old.per_product[element.new.product].total, 0) + element.new.price
              }
          }) : {
              "${element.new.product}" = {
                  total = element.new.price
              }
          }
    }
  }
}
```

if `element.old.per_product` already exists, it merges it but provides a new object for the current `element.new.product` using the old value (if any) to recompute the total. But the merge is overwriting old keys...
Any idea why?

Reply:
as both the outer and inner map blocks iterate over `orders`, I don't know which iterator names to give them.. ideas? @mariux

Reply:
inner and outer? ;)

Reply:
or `per_name` and `per_product` ?